analysis/quantifier-comparisons/03-memory_time_comparison.Rmd

```diff
@@ -0,0 +1,287 @@
+  geom_point(size = 2) + 
+  theme_classic() + 
+  xlab("") +
+  ylab("Run Time (Minutes)")
+```
+
+```{r}
+# another way to visualize it if we want to group all the samples together
+summary_stats_df %>%
+  mutate(Tool = fct_reorder(Tool, total_time, .fun = 'median')) %>%
+  ggplot(aes(x = reorder(Tool, total_time), y = total_time, fill = Tool)) + 
+  geom_boxplot() + 
+  theme_classic() +
+  xlab("") +
+  ylab("Run Time (Minutes)")
+```
+
+
+```{r}
+ggplot(summary_stats_df, aes(x = sample_id, y = total_rss, color = Tool)) + 
+  geom_point(size = 2) + 
+  theme_classic() + 
+  xlab("") +
+  ylab("Memory (GB)")
+```
+
+```{r}
+# another way to visualize it if we want to group all the samples together
+summary_stats_df %>%
+  mutate(Tool = fct_reorder(Tool, total_rss, .fun = 'median')) %>%
+  ggplot(aes(x = reorder(Tool, total_rss), y = total_rss, fill = Tool)) +
+  geom_boxplot() + 
+  theme_classic() +
+  xlab("") +
+  ylab("Memory (GB)")
+```
+
+
+After looking at the runtime and memory used independently of any other 
+observations, we can see that Kallisto is consistently the fastest, although 
+alevin-fry in --sketch mode is also quite fast. This is probably because it is
+supposed to mimic kallisto. Cellranger is consistently the slowest by almost 
+3x more than the next fastest, Alevin/Alevin-fry without --sketch mode which 
+are both somewhere in the middle.
+
+In terms of memory, Alevin-fry with/without --sketch uses the least amount
+of memory using < 5 total GB of memory on all 4 samples. Kallisto is also 
+consistently using ~ 11 GB of memory for all 4 samples. Alevin and cellranger
+are higher than this, with Alevin requiring the most amount of memory closer
+to ~ 20 GB of memory per sample. 
+
+Next, let's see if the number of reads in each fastq file impacts the runtime
+and/or memory usage. 
+
+## Time and Memory vs. # of Reads 
+
+```{r}
+# run time vs. # of fastq reads
+ggplot(summary_stats_df, aes(x = Reads, y = total_time, color = Tool)) +
+  geom_point(size = 2) + 
+  geom_line() +
+  theme_classic() + 
+  xlab("Number of Reads") +
+  ylab("Run Time (Minutes)")
+```
+
+
+```{r}
+ggplot(summary_stats_df, aes(x = Reads, y = total_rss, color = Tool)) + 
+  geom_point(size = 2) + 
+  geom_line() +
+  theme_classic() + 
+  xlab("Number of Reads") +
+  ylab("Memory (GB)")
+```
+
+From this, it looks like there is no effect on the number of reads found in 
+the fastq files to the memory that is required for any of the processes, except
+maybe with alevin-fry both with/without --sketch. It appears that the memory
+usage almost doubles from ~2-3 GB to closer to ~5-6 GB. Even though memory
+usage doesn't appear to be that affected, run time does appear to be affected
+by the number of reads in the sample if running on cellranger. As the number
+of reads increase, there is a sharp increase in the run time for cellranger, 
+where as that is not the case for the other tools.
```

Comment:
I think there may be an issue of scaling here, where the increases for alevin-fry and kallisto are there, but not as large. I also wonder if something about the machines chosen by AWS might have varied for Alevin/Alevin-fry in the first case. Seems odd that there wouldn't be an effect of the number of reads on run time, but maybe something else about the data is affecting runtimes.

Reply:
Yes, it does look like the number of reads may also affect Alevin-fry, but the scales make it less visually obvious. I also think there might be clearer trends with more samples since there are only 4 samples here.